### PR TITLE
Adding tooltip

### PR DESCRIPTION
**Pull request by kdreese**
And rewriting the way the entire grid interacts to get there

### CODE REVIEW

File: src/cards/actions/create_ranged_unit.gd [hunk 1]
```diff
@@ -2,16 +2,18 @@
 extends CardAction
 
 
-func perform_action(data: CardData, lane: int) -> void:
+func can_perform(_data: CardData, grid_pos: Vector2i, is_enemy: bool) -> bool:
+	return grid_pos.x > 7 and (grid_pos.y < 3 if is_enemy else grid_pos.y > 2)
+
```
Comment (admoore0) [inline src/cards/actions/create_ranged_unit.gd:7]: only one newline, literally unplayable

File: src/cards/actions/increase_health.gd [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
+func perform_action(data: CardData, grid_pos: Vector2i) -> void:
 	if data.extra_data.size() != 1:
 		push_error("increase_health expects 1 argument of the health to increase")
 		return
 	var health_bonus := int(data.extra_data[0])
 	var health_bar
-	if lane < 3: # The enemy is using this
+	if grid_pos.y < 3: # The enemy is using this
 		health_bar = game.red_castle_health_bar
 		game.get_node("RightHealSound").play()
```
Comment (admoore0) [inline src/cards/actions/increase_health.gd]: Let's add is_enemy to perform_action as well. Saves some hassle

File: src/states/game/game.gd [hunk 3]
```diff
@@ -35,17 +33,15 @@ const COPY_ROUND_DOWNTIME = 3
...
+@onready var drop_points: Node2D = %DropPoints
+@onready var blue_castle_door: Vector2 = $DefenseBridgePoint.position
+@onready var red_castle_door: Vector2 = $AttackBridgePoint.position
 
 @onready var win_sound: AudioStreamPlayer = $WinSound
 @onready var draw_sound: AudioStreamPlayer = $DrawSound
 
 
-
+var drop_world_pos: Dictionary
```
Comment (admoore0) [inline src/states/game/game.gd]: Not a fan of the name. Perhaps `grid_to_world_pos`?

File: src/units/unit.gd [hunk 4]
```diff
@@ -2,3 +2,5 @@ class_name Unit
 extends Node2D
 
 var grid_position: Vector2i
+var world_position_dict : Dictionary
```
Comment (admoore0) [inline src/units/unit.gd]: Probably rename this to `grid_to_world_position` too

**Review by admoore0 — CHANGES_REQUESTED** [review]

File: src/cards/actions/increase_health.gd (hunk 2 above)
Comment (admoore0) [reply src/cards/actions/increase_health.gd]: Basically all the places where you use grid_position.y can use that instead.

File: src/cards/actions/card_action.gd [hunk 5]
```diff
@@ -15,10 +15,10 @@ func set_game(game_scene: Node) -> void:
...
 
 ## An optional function which can prevent an action from being run (and card consumed).
-func can_perform(_data: CardData, _lane: int) -> bool:
+func can_perform(_data: CardData, _grid_pos: Vector2i, _is_enemy: bool) -> bool:
 	return true
 
 
 ## The action which is performed when the card is dropped. Accepts the card data and lane
-func perform_action(_data: CardData, _lane: int) -> void:
+func perform_action(_data: CardData, _grid_pos: Vector2i) -> void:
```
Comment (thearst3rd) [inline src/cards/actions/card_action.gd]: Regarding @admoore0's comment: Here is where we should add an `_is_enemy: bool` property, and add it to the implementations of `perform_action` as well

File: src/states/game/game.gd (hunk 3 above)
Comment (thearst3rd) [reply src/states/game/game.gd]: GDScript doesn't support strict typing of dictionaries but I'd put in a comment detailing the actual typing:

```gdscript
var grid_to_world_pos: Dictionary # Dictionary[Vector2i, Vector2]
```

File: src/ui/tooltip.gd [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+@onready var info_text: Label = %InfoText
+@onready var damage_amt: Label = %DamageAmt
+@onready var health_amt: Label = %HealthAmt
+@onready var move_amt: Label = %MoveAmt
+@onready var damage_icon: TextureRect = %DamageIcon
+@onready var health_icon: TextureRect = %HealthIcon
+@onready var move_icon: TextureRect = %MoveIcon
+
+
+func initialize(unit : Unit) -> void:
```
Comment (thearst3rd) [inline src/ui/tooltip.gd]: Super pedantic:
```gdscript
func initialize(unit: Unit) -> void:
```
(no space in between `unit` and `:`, I am very sorry)

**Review by thearst3rd — CHANGES_REQUESTED** [review]

File: src/ui/tooltip.gd (hunk 6 above)
Comment (kdreese) [reply src/ui/tooltip.gd]: image

**Review by admoore0 — APPROVED** [review]

**Review by thearst3rd — APPROVED** [review]

**Pull request merged**